### PR TITLE
Directus and IoT: Sensor Data with an ESP32

### LINKED ISSUE #119: Directus and IoT: Sensor Data with an ESP32
### What is your idea?

This is an article that gets users started with using Directus for IoT projects. It’s a simple project that demonstrates how to collect and visualize temperature and humidity data from a DHT22 temperature and humidity sensor and send this data to Directus through an ESP32. An ESP32 is a board that allows IoT developers and hobbyists to work with sensors and controllers with built-in Wifi support. This Wifi support allows the board to connect to the internet and send data to destinations such as Directus by making HTTP calls. Below is a quick overview of the major sections of the article:

1. It begins by highlighting the necessity of collecting and visualizing temperature and humidity data in various settings, including industrial and home environments.
2. Next, it illustrates both physical and simulated circuit configurations to aid readers in implementing the project using either physical components or the Wowki IoT simulator.
3. The setup process for Directus is then outlined, offering guidance on configuring it either through Directus Cloud or a self-hosted setup.
4. Instructions are provided for creating a collection named "temperature_and_humidity," designed to store temperature and humidity data with four essential fields: id, temperature, humidity, and date_created.
5. A demonstration follows on how to send data to this collection using the Directus REST API.
6. Subsequently, readers are acquainted with making HTTP calls from an ESP32 using embedded C++ scripting.
7. The article proceeds to showcase a demo of the HTTP POST request responsible for sending data to the "temperature_and_humidity" collection from the ESP32 script.
8. The final step involves visualizing the accumulated temperature and humidity data using Directus Dashboards.
9. Concluding the article, a discussion is presented, reflecting on the completed tasks and potential enhancements, such as incorporating a board_id field to identify the source board for each record within the collection.

### What are the key takeaways from your post?

1. How to store data from an IoT system to the cloud with the aid of Directus
2. How to use Directus dashboards to easily visualize data from a database
3. How to use the Directus API to make POST requests
4. How to effectively utilize role permissions and access tokens in a Directus-powered project

### What type of post is this?

Full Project Build with Directus

### Country of residence

 Nigeria

### Terms & Conditions

- [X] You understand that submitting this form does not mean we accept your proposal. We will accept proposals at our discretion.
- [X] You understand that to be paid you will need to generate an invoice to us once the final draft of your work is completed.
- [X] You understand that you will grant us an unlimited license to post and promote your work as part of this program. You may repost, but using our copy as the canonical source.
- [X] You are aged 18 or over.

### PR DESCRIPTION
**Pull request by vicradon**
Completed the first draft of the article on Directus and IoT: Sensor Data with an ESP32.

Closes #119

### CODE REVIEW

File: directus-and-iot-sensor-data-with-an-esp32/index.md [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+3. Y-Axis - Humidity
+4. Y-Axis Function - Max
+5. Value Decimals - 2
+6. Color - #E35168
+7. Panel header title - Humidity over time
+8. Panel header icon - Water DO
+
+You should have the temperature and humidity charts displaying how your values changed over time. You can conclude this section by saving this dashboard.
+
+![temperature and humidity trends over time](./temperature_and_humidity_trends_over_time.png)
```
Comment (phazonoverload) [inline directus-and-iot-sensor-data-with-an-esp32/index.md:316]:
```suggestion
## Visualizing Data in Directus Insights

Directus Insights allows you to create multiple panels in a dashboard, powered by data in your project.

To show the change over time for temperature, create a bar chart with the following settings:

1. Collection - Temperature and Humidity
2. X-Axis - Date Created
3. Y-Axis - Temperature
4. Y-Axis Function - Max
5. Value Decimals - 2
6. Color - #E35168

You can repeat this for humidity, and any other data inside of your project. 

![temperature and humidity trends over time](./temperature_and_humidity_trends_over_time.png)
```

File: directus-and-iot-sensor-data-with-an-esp32/index.md [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+## Conclusion
+
+In this tutorial, you learned how to collect temperature and humidity data from a DHT22 sensor and log it to a database with the aid of Directus. You learned how to visualize how this data changes over time using Directus dashboards.
+
+Directus presents a complete BaaS solution allowing you to build content-focused and traditional web applications. It offers a plethora of services including database mirroring, user authentication, OAuth2, and HTTP APIs over data (REST and GraphQL). You can self-host Directus or use their cloud API with reasonable pricing. [Try Directus today](https://directus.cloud/)!
```
Comment (phazonoverload) [inline directus-and-iot-sensor-data-with-an-esp32/index.md:322]:
```suggestion
## Summary

In this tutorial, you learned how to collect temperature and humidity data from a DHT22 sensor and log it to a Directus project. You also learned how to visualize how this data changes over time using Directus Insights.
```

**Review by phazonoverload — COMMENTED** [review]
Good post. You'll notice most of my suggestions are removing a lot of content. 

You don't need to teach people how to use Directus - for this novel project it's ok to assume they hold that knowledge, or link out to documents like the quickstart when they don't. 

You'll also notice that I removed your thoughts about how to host Directus - it's better to stay generic and link to a quickstart. This isn't the place for that editorial voice. 

But very good - take a look through suggestions / requests / comments and re-request a review once you've made changes.

**Comment by phazonoverload** [comment]
Hey @vicradon how are you getting on? I'm away this week but would love to be able to review changes next Monday if possible. :)

File: directus-and-iot-sensor-data-with-an-esp32/index.md [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+A DHT22 sensor can connect directly to an ESP32 using three pins. DHT22 comes in two types, 3-pin type and 4-pin type. The 3-pin type doesn't require extra configuration. You connect ground to ground, VCC to 5V output, and data to a GPIO pin, say pin 13. For the 4-pin type, ignore the 3rd pin from the left and connect as shown in the image below:
+
+![DHT22 to ESP32](./DHT22_to_ESP32.png)
+
+### Installing DHT22 Sensor Library
+
+Since you will program your ESP32 using the Arduino IDE, you must install the [DHT sensor library by Adafruit](https://www.arduino.cc/reference/en/libraries/dht-sensor-library/). Search for the "DHT Sensor Library" in your library manager and install the corresponding library authored by Adafruit. Use the image below as a reference.
+
+![Installing the DHT22 sensor library](./Installing_the_DHT22_sensor_library.png)
```
[the remark this answers is not in the document]
Comment (vicradon) [reply directus-and-iot-sensor-data-with-an-esp32/index.md:73]: Done

File: directus-and-iot-sensor-data-with-an-esp32/index.md [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+```
+
+After running the curl command above, you will see a new value in the collection on your Directus dashboard (/admin/content/temperature_and_humidity).
+
+You can run the cURL command with different values of temperature and humidity to see more data in your collection, but first, take a moment to look at the URL used in the cURL command above. It is your Directus root URL at localhost:8055 with a path /items/temperature_and_humidity. Directus gives a straightforward way to interact with collections via a REST API by appending /items to the root URL and then the collection name. You can perform API operations using the standard REST principles such as:
+
+- GET /items/temperature_and_humidity
+- POST /items/temperature_and_humidity
+- PATCH /items/temperature_and_humidity
+- DELETE /items/temperature_and_humidity
```
[the remark this answers is not in the document]
Comment (vicradon) [reply directus-and-iot-sensor-data-with-an-esp32/index.md:61]: Done

File: directus-and-iot-sensor-data-with-an-esp32/index.md [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+Since you will program your ESP32 using the Arduino IDE, you must install the [DHT sensor library by Adafruit](https://www.arduino.cc/reference/en/libraries/dht-sensor-library/). Search for the "DHT Sensor Library" in your library manager and install the corresponding library authored by Adafruit. Use the image below as a reference.
+
+![Installing the DHT22 sensor library](./Installing_the_DHT22_sensor_library.png)
+
+### Connecting the ESP32 Board to your Computer
+
+You can see the values from the DHT22 sensor in the Arduino serial monitor. After connecting your ESP32 to your computer, choose a board and port that corresponds to your purchased board and available port on your computer.
+
+![Board selector page](./Board_selector_page.png)
```
[the remark this answers is not in the document]
Comment (vicradon) [reply directus-and-iot-sensor-data-with-an-esp32/index.md:79]: Done

File: directus-and-iot-sensor-data-with-an-esp32/index.md [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+1. Your WiFI SSID, i.e. the name of your WiFi network, as the value of the `ssid` variable on line 6.
+2. Your WiFi password on line 7.
+3. Your Directus esp32-board user token that you used with cURL at the `Creating dummy temperature and humidity values` section of this article. You can still [regenerate the token if you lost it](./regenerate_lost_token.png). If you have the token, set it as the <TOKEN> placeholder value on line 8.
+4. Your WiFi gateway address is defined on line 9 as `directusEndpoint`. It won't start with localhost because the ESP32 is running as a separate system. So you must check the gateway address of your local network. This address will either start with 10, 172, or 192. Follow [this blog post](https://nordvpn.com/blog/find-router-ip-address/) for instructions for checking your gateway address on different operating systems. Note that your ESP32 and your computer running Directus must be connected to the same WiFi network for this to work.
+
+With all the changes made, you can upload your script to your ESP32 and observe it log temperature and humidity data on your Directus `temperature_and_humidity` collection.
```
[the remark this answers is not in the document]
Comment (vicradon) [reply directus-and-iot-sensor-data-with-an-esp32/index.md]: Done

**Comment by vicradon** [comment]
Hi @phazonoverload 
Thank you for reviewing my article. I have effected the changes you requested.

**Comment by phazonoverload** [comment]
Thank you for writing this, and we are happy to consider it done in it's current state. My team will schedule it in to be published. 

This repo is just for authoring/reviewing - the actual piece will be published via a Directus project, so we'll take it from here. I'll close this issue to keep the main branch clean, but this is a acceptance of your work.

Could you email me at [devrel@directus.io](mailto:devrel@directus.io) and we'll share what we need to get you paid. Thanks!

**Pull request closed without merging**